### ISSUE TITLE
jl_deserialize_value crash during Pkg.add("MatlabCompat")

### ISSUE THREAD

**Issue by tkelman** [JuliaLang/julia#20419]
Bisect points to c99e12c9338 as introducing this, cc @simonbyrne
```
~/Julia/julia-0.6$ ./julia -e 'versioninfo()'
Julia Version 0.6.0-dev.2540
Commit ad5cd7b* (2017-02-02 21:32 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E3-1241 v3 @ 3.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
~/Julia/julia-0.6$ rm -rf jlpkgtmp
~/Julia/julia-0.6$ JULIA_PKGDIR=$PWD/jlpkgtmp ./julia -e 'Pkg.init(); Pkg.add("MatlabCompat")'
INFO: Initializing package repository /home/tkelman/Julia/julia-0.6/jlpkgtmp/v0.6
INFO: Cloning METADATA from https://github.com/JuliaLang/METADATA.jl
INFO: Cloning cache of AxisArrays from https://github.com/JuliaArrays/AxisArrays.jl.git
<etc>
INFO: Cloning cache of Zlib from https://github.com/dcjones/Zlib.jl.git
INFO: Installing AxisArrays v0.0.3
<etc>
INFO: Installing Zlib v0.1.12

signal (15): Terminated
while loading no file, in expression starting on line 0
unknown function (ip: 0x7f2ecc81aeed)
memcpy at /usr/include/x86_64-linux-gnu/bits/string3.h:51 [inlined]
_ios_read at /home/tkelman/Julia/julia-0.6/src/support/ios.c:275 [inlined]
ios_read at /home/tkelman/Julia/julia-0.6/src/support/ios.c:331
jl_deserialize_value_ at /home/tkelman/Julia/julia-0.6/src/dump.c:2057
jl_deserialize_value_svec at /home/tkelman/Julia/julia-0.6/src/dump.c:1534 [inlined]
jl_deserialize_value_ at /home/tkelman/Julia/julia-0.6/src/dump.c:1979
jl_deserialize_struct at /home/tkelman/Julia/julia-0.6/src/dump.c:1880
jl_deserialize_value_any at /home/tkelman/Julia/julia-0.6/src/dump.c:1969 [inlined]
jl_deserialize_value_ at /home/tkelman/Julia/julia-0.6/src/dump.c:2065
jl_deserialize_value at /home/tkelman/Julia/julia-0.6/src/dump.c:1517 [inlined]
jl_deserialize_value_array at /home/tkelman/Julia/julia-0.6/src/dump.c:1595
<etc>
```
full output: https://gist.github.com/5ea4ace3667e6043e9c8ec87fbe021de

**Comment by simonbyrne** [comment]
Interesting, on OS X I get a bit more output at the bottom:
```
signal (15): Terminated: 15
while loading no file, in expression starting on line 0
_platform_memmove$VARIANT$Haswell at /usr/lib/system/libsystem_platform.dylib (unknown line)
unknown function (ip: 0xffffffffffffffff)
Allocations: 117982 (Pool: 117810; Big: 172); GC: 0
ERROR: GitError(Code:ENOTFOUND, Class:Tree, the path 'REQUIRE' does not exist in the given tree)
 in macro expansion at ./libgit2/error.jl:99 [inlined]
 in Base.LibGit2.GitUnknownObject(::Base.LibGit2.GitRepo, ::String) at ./libgit2/repository.jl:100
 in revparseid(::Base.LibGit2.GitRepo, ::String) at ./libgit2/repository.jl:128
 in (::Base.Pkg.Read.##5#6{String})(::Base.LibGit2.GitRepo) at ./pkg/read.jl:185
 in with(::Base.Pkg.Read.##5#6{String}, ::Base.LibGit2.GitRepo) at ./libgit2/types.jl:577
 in requires_path(::String, ::Dict{VersionNumber,Base.Pkg.Types.Available}) at ./pkg/read.jl:182
 in requires_list at ./pkg/read.jl:196 [inlined] (repeats 2 times)
 in build!(::Array{String,1}, ::Pipe, ::Set{Any}) at ./pkg/entry.jl:588
 in build!(::Array{String,1}, ::Dict{Any,Any}, ::Set{Any}) at ./pkg/entry.jl:634
 in build(::Array{String,1}) at ./pkg/entry.jl:656
 in resolve(::Dict{String,Base.Pkg.Types.VersionSet}, ::Dict{String,Dict{VersionNumber,Base.Pkg.Types.Available}}, ::Dict{String,Tuple{VersionNumber,Bool}}, ::Dict{String,Base.Pkg.Types.Fixed}, ::Dict{String,VersionNumber}, ::Set{String}) at ./pkg/entry.jl:571
 in edit(::Function, ::String, ::Base.Pkg.Types.VersionSet, ::Vararg{Base.Pkg.Types.VersionSet,N} where N) at ./pkg/entry.jl:30
 in (::Base.Pkg.Entry.##1#3{String,Base.Pkg.Types.VersionSet})() at ./task.jl:335
Stacktrace:
 [1] sync_end() at ./task.jl:287
 [2] macro expansion at ./task.jl:303 [inlined]
 [3] add(::String, ::Base.Pkg.Types.VersionSet) at ./pkg/entry.jl:50
 [4] (::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#add,Tuple{String}})() at ./pkg/dir.jl:31
 [5] cd(::Base.Pkg.Dir.##2#3{Array{Any,1},Base.Pkg.Entry.#add,Tuple{String}}, ::String) at ./file.jl:69
 [6] #cd#1(::Array{Any,1}, ::Function, ::Function, ::String, ::Vararg{Any,N} where N) at ./pkg/dir.jl:31
 [7] add(::String) at ./pkg/pkg.jl:102
```

Running
```
├ find .julia/v0.6 -mindepth 1 -maxdepth 1 -type d '!' -exec test -e "{}/REQUIRE" ';' -print
.julia/v0.6/Color
.julia/v0.6/METADATA
```

For some reason, [Color.jl v0.2.11](https://github.com/JuliaArchive/Color.jl/releases/tag/v0.2.11) from 2014 is getting installed, which lacks a REQUIRE file.

**Comment by simonbyrne** [comment]
Some dependency must have an upper bound constraint, as MatlabCompat v0.1.2 is being installed, which has Color.jl in [its REQUIRE file](https://github.com/MatlabCompat/MatlabCompat.jl/blob/v0.1.2/REQUIRE)

**Comment by simonbyrne** [comment]
I've fixed the immediate cause of the error in #20433.

**Comment by tkelman** [comment]
nothing funny about upper bounds, that package just hasn't been tagged in ages. shouldn't cause a crash if you try to add it though.

**Comment by tkelman** [comment]
ah no there is an incomplete bounds issue where later versions of Color have upper bounds on FixedPointNumbers but the early versions didn't depend on it, so Pkg is using an old Color in order to use a newer FixedPointNumbers.

**Comment by simonbyrne** [comment]
Okay, so there are a couple of problems, which we should address:
1. We're getting a crash rather than a clean error. Not sure why this is happening: is this something specific about Pkg?
2. Pkg is throwing an error (fixed by  #20433).
3. Pkg perhaps should warn when installing non-latest versions

**Comment by tkelman** [comment]
1 may have been a segfault or some other kind of data corruption from libgit2?

yes to 3, we should also make Pkg.status say something when a version is held back to something less than the latest available for the running Julia version
